service/discovery.go

```diff
@@ -0,0 +1,75 @@
+package service
+
+import (
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/hashicorp/consul/api"
+	"golang.org/x/sync/syncmap"
+)
+
+// DefaultCacheInvalidationTimeout for updating instances
+const (
+	DefaultCacheInvalidationTimeout = 30
+)
+
+// Services for discovery service
+type Services struct {
+	ConsulClient             *api.Client
+	Instances                *syncmap.Map
+	CacheInvalidationTimeout int64
+}
+
+// GetEndpoint by service name
+func (s *Services) GetEndpoint(serviceName string) (url *url.URL, err error) {
+	resolver := s.UpdateInstences(serviceName)
+	if resolver != nil && len(resolver.endpoints) > 0 {
+		url = resolver.getHealthyInstanceEndpoint()
+	} else {
+		err = fmt.Errorf("no registered or healtly instances for service: %s", serviceName)
+	}
+	return
+}
+
+// UpdateInstences get service instances from service discovery
+func (s *Services) UpdateInstences(serviceName string) (resolver *Resolver) {
+	value, instancesExists := s.Instances.Load(serviceName)
+	if instancesExists {
+		resolver = value.(*Resolver)
+	} else {
+		s.Instances = new(syncmap.Map)
+		resolver = NewResolver(s.ConsulClient)
+		resolver.updateLastTimestamp()
+	}
+
+	if !instancesExists || time.Now().Unix()-resolver.TTL >= s.CacheInvalidationTimeout {
+		if instancesExists {
```

Comment:
I can't reproducing this problem in practice. I do test with 100 goroutines and 50 service instances in consul with updating cache/instances time in two cases: 1, 10 seconds. Test time ~4h and I can't see cases with updating cache twice.

Reply:
Łukasz is right. I accept the fact it may happen on Akubra startup, when cache is empty, but double (or, in fact, triple...) update should never happen when cache is already filled with instance data.